core/types/rollup_l1_cost.go

```diff
@@ -1,83 +0,0 @@
-// Copyright 2022 The go-ethereum Authors
-// This file is part of the go-ethereum library.
-//
-// The go-ethereum library is free software: you can redistribute it and/or modify
-// it under the terms of the GNU Lesser General Public License as published by
-// the Free Software Foundation, either version 3 of the License, or
-// (at your option) any later version.
-//
-// The go-ethereum library is distributed in the hope that it will be useful,
-// but WITHOUT ANY WARRANTY; without even the implied warranty of
-// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
-// GNU Lesser General Public License for more details.
-//
-// You should have received a copy of the GNU Lesser General Public License
-// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
-
-package types
-
-import (
-	"math/big"
-
-	"github.com/ethereum/go-ethereum/common"
-	"github.com/ethereum/go-ethereum/params"
-)
-
-type RollupGasData struct {
-	Zeroes, Ones uint64
-}
-
-func (r RollupGasData) DataGas(time uint64, cfg *params.ChainConfig) (gas uint64) {
-	gas = r.Zeroes * params.TxDataZeroGas
-	if cfg.IsRegolith(time) {
-		gas += r.Ones * params.TxDataNonZeroGasEIP2028
-	} else {
-		gas += (r.Ones + 68) * params.TxDataNonZeroGasEIP2028
-	}
-	return gas
-}
-
-type StateGetter interface {
-	GetState(common.Address, common.Hash) common.Hash
-}
-
-// L1CostFunc is used in the state transition to determine the cost of a rollup message.
-// Returns nil if there is no cost.
-type L1CostFunc func(blockNum uint64, blockTime uint64, dataGas RollupGasData, isDepositTx bool) *big.Int
-
-var (
-	L1BaseFeeSlot = common.BigToHash(big.NewInt(1))
-	OverheadSlot  = common.BigToHash(big.NewInt(5))
-	ScalarSlot    = common.BigToHash(big.NewInt(6))
-)
-
-var L1BlockAddr = common.HexToAddress("0x4200000000000000000000000000000000000015")
-
-// NewL1CostFunc returns a function used for calculating L1 fee cost.
-// This depends on the oracles because gas costs can change over time.
-// It returns nil if there is no applicable cost function.
-func NewL1CostFunc(config *params.ChainConfig, statedb StateGetter) L1CostFunc {
-	cacheBlockNum := ^uint64(0)
-	var l1BaseFee, overhead, scalar *big.Int
-	return func(blockNum uint64, blockTime uint64, dataGas RollupGasData, isDepositTx bool) *big.Int {
-		rollupDataGas := dataGas.DataGas(blockTime, config) // Only fake txs for RPC view-calls are 0.
-		if config.Optimism == nil || isDepositTx || rollupDataGas == 0 {
-			return nil
-		}
-		if blockNum != cacheBlockNum {
-			l1BaseFee = statedb.GetState(L1BlockAddr, L1BaseFeeSlot).Big()
-			overhead = statedb.GetState(L1BlockAddr, OverheadSlot).Big()
-			scalar = statedb.GetState(L1BlockAddr, ScalarSlot).Big()
-			cacheBlockNum = blockNum
-		}
-		return L1Cost(rollupDataGas, l1BaseFee, overhead, scalar)
-	}
```

Comment:
Critical difference: the state lookups used to be at evaluation-time of the function.

If a `L1CostFunc` is created at the start of the block, then the transactions after the L1 Block info tx is applied all read the most up-to-date values.
Caching outside the function makes sense, but is a breaking change, as in a shared block-context it would result in outdated lookups, if not explicitly invalidated with this block-num check.

The block-num check and caching here was removed in the new version. This should be introduced back, to avoid critical chain-split.

Reply:
Updated to require caller pass in the blockTime and look up values dynamically on first call or when a different blockTime is passed in. Clarified in the docs why this delayed-lookup behavior is consensus critical.